pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,254 @@
+
+	return types.PatchEnvelopeInfoList{
+		Envelopes: res,
+	}
+}
+
+func (pes *PatchEnvelopes) processMessages() {
+	for {
+		select {
+		case volumeStatus := <-pes.VolumeStatusCh:
+			pes.updateVolumeStatus(volumeStatus)
+		case newPatchEnvelopeInfo := <-pes.PatchEnvelopeInfoCh:
+			pes.updateEnvelopes(newPatchEnvelopeInfo)
+		}
+	}
+}
+
+func (pes *PatchEnvelopes) updateVolumeStatus(volumeStatus PatchEnvelopesVsCh) {
+	switch volumeStatus.Action {
+	case PatchEnvelopesVsChActionPut:
+		pes.updateExternalPatches(volumeStatus.Vs)
+	case PatchEnvelopesVsChActionDelete:
+		pes.deleteExternalPatches(volumeStatus.Vs)
+	}
+}
+
+func (pes *PatchEnvelopes) updateExternalPatches(vs types.VolumeStatus) {
+	if vs.State < types.CREATED_VOLUME {
+		return
+	}
+
+	pes.Lock()
+	defer pes.Unlock()
+
+	volumeExists := false
+	for i := range pes.completedVolumes {
+		if pes.completedVolumes[i].VolumeID == vs.VolumeID {
+			pes.completedVolumes[i] = vs
+			volumeExists = true
+			break
+		}
+	}
+	if !volumeExists {
+		pes.completedVolumes = append(pes.completedVolumes, vs)
+	}
+
+	if err := pes.processVolumeStatus(vs); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) deleteExternalPatches(vs types.VolumeStatus) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	i := 0
+	for _, vol := range pes.completedVolumes {
+		if vol.VolumeID == vs.VolumeID {
+			break
+		}
+		i++
+	}
+
+	pes.completedVolumes[i] = pes.completedVolumes[len(pes.completedVolumes)-1]
+	pes.completedVolumes = pes.completedVolumes[:len(pes.completedVolumes)-1]
+}
+
+func (pes *PatchEnvelopes) updateEnvelopes(peInfo []types.PatchEnvelopeInfo) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	pes.envelopes = peInfo
+	if err := pes.processVolumeStatusList(pes.completedVolumes); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) processVolumeStatus(vs types.VolumeStatus) error {
+	for i := range pes.envelopes {
+		for j := range pes.envelopes[i].VolumeRefs {
+			volUUID, err := uuid.FromString(pes.envelopes[i].VolumeRefs[j].ImageID)
+			if err != nil {
+				return err
+			}
+			if volUUID == vs.VolumeID {
+				blob, err := blobFromVolRef(pes.envelopes[i].VolumeRefs[j], vs)
+				if err != nil {
+					return err
+				}
+
+				if idx := types.CompletedBinaryBlobIdxByName(pes.envelopes[i].BinaryBlobs, blob.FileName); idx != -1 {
+					pes.envelopes[i].BinaryBlobs[idx] = *blob
+				} else {
+					pes.envelopes[i].BinaryBlobs = append(pes.envelopes[i].BinaryBlobs, *blob)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (pes *PatchEnvelopes) processVolumeStatusList(volumeStatuses []types.VolumeStatus) error {
+	for _, vs := range volumeStatuses {
+		if err := pes.processVolumeStatus(vs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func blobFromVolRef(vr types.BinaryBlobVolumeRef, vs types.VolumeStatus) (*types.BinaryBlobCompleted, error) {
+	sha, err := fileutils.ComputeShaFile(vs.FileLocation)
```

Comment:
Wait, this call can take minutes for a huge volume and a slow CPU. While it can't deadlock, it means that it can cause a watchdog due to the caller not reading the channels in a timely manner.

Reply:
Note that where we use this in pkg/pillar/cmd/verifier/verifier.go we are not subject to a watchdog check (since it is not running in the main verifier event loop but in a go routine processing a single VerifyImageConfig). But here due to the channel send in the Handle functions you can indirectly block the main event loop in zedrouter.

Reply:
What if we wrap sending information to a channel in a go routine in handlers as well?
```golang
func (z *zedrouter) handlePatchEnvelopeModify(//...
  //...
  go func() {
    z.patchEnvelopes.PatchEnvelopeInfoCh <- peInfo.Envelopes
  }
}
```
That way handlers will not be hanged for a long time so we will not cause watchdog. 

I tried to implement delayed update, when client app instance really needs it (i.e. when PatchEnvelope.Get() is called), but that won't solve the problem, because we will block writing to channels during Get() because of the mutex.

Alternative to using goroutines will be to change signature and call updateVolumeStatus and updateEnvelopes directly. It'll still block handle, but it will not trigger watchdog. I think goroutine option makes more sense.

Reply:
What about using queue (of unlimited size) for collecting work and processing it from a separate go routine running inside PatchEnvelopes?

Reply:
But when would you want to read data? Will Get be added to queue to see intermediate state? And it could be that one app instance is blocked because we are processing patch envelopes for another app instance (if we won't queue Get() and will wait for all of it to be done).

Reply:
@eriknordmark I tried to wrap sending to channels to goroutines and eve updates and patch envelopes endpoints are accessible from app instances.

Reply:
The goroutine approach also consumes unbounded memory just like an unbounded channel; you can have tons of goroutines sitting around blocked to send. 

The only pattern I know which avoids this is to avoid queuing events but instead updating the intended state and then queing a single notification that something has changed using a nonblocking send on a notification channel.

Reply:
@eriknordmark I think that new version addresses your concerns. It's not exactly sending notification for a channel, but rather we change or add VolumeStatus`es and PatchEnvelopeInfo in goroutine and lock it only during appending to channels. In case of Get() we lock the structure only for copying VolumeStatus and PatchEnvelopeInfo slices and then process them to give it to app instance only if there is a change. That way we don't block sending for sha computation and whole iteration, but rather just for copying two slices.
I left wrapping goroutines as a separate commit for now, to discuss weather or not we need it, because it'll help us avoid watchdog, but as you said it'll potentially create unbounded queue. And last thing is debatable since blocking will be done only for copying two slices, assigning one slice to another and iterating over VolumeStatus slice.